---
title: "Quality Control (QC) and Exploration of scRNA-seq Datasets"
output: rmarkdown::github_document
date: "2024-09-02"
---

Note: There is no standard method for performing scRNA-seq. QC is mostly done by looking for cells that are outliers compared to others in the dataset, i.e., there are no external/independent quality standards. Should be careful and thoughtful when comparing quality metrics across datasets collected/sequenced using different protocols.

### Dataset Construction and QC

Load libraries. Use `suppressPackageStartupMessages` so the start up messages don't show up.

```{r}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(AnnotationDbi))
suppressPackageStartupMessages(library(org.Hs.eg.db))
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
```

Load data.

```{r}
molecules <- read.delim("course/data/tung/molecules.txt", row.names = 1)
annotation <- read.delim("course/data/tung/annotation.txt",stringsAsFactors = T)
```

Take a look at the data.

```{r}
head(molecules[,1:3])
```

```{r}
head(annotation)
```

In this dataset, they used both unique molecular identifiers (UMIs) and ERCC spike-ins. We use `altExp` (alternative Experiment) to separate the spike-ins from the main dataset.

```{r}
umi <- SingleCellExperiment(assays = list(counts = as.matrix(molecules)), colData = annotation)
altExp(umi, "ERCC") <- umi[grep("^ERCC-", rownames(umi)), ]
umi <- umi[grep("^ERCC-", rownames(umi), invert = T), ]
```

Map ENSEMBL IDs to gene symbols.

```{r}
gene_names <- mapIds(org.Hs.eg.db, keys=rownames(umi), keytype="ENSEMBL", columns="SYMBOL", column="SYMBOL")
```

mapIds returns one symbol per ID. Using `table` command, find out how many genes were not annotated.

```{r}
rowData(umi)$SYMBOL <- gene_names
table(is.na(gene_names))
```

Remove genes where no symbols matched.

```{r}
umi <- umi[! is.na(rowData(umi)$SYMBOL), ]
```

Check if there are any mitochondrial genes in the newly annotated symbols.

```{r}
grep("^MT-", rowData(umi)$SYMBOL, value = T)
```

Check for ribosomal proteins (starting with "RPL" or "RPS")

```{r}
grep("^RP[LS]",rowData(umi)$SYMBOL,value = T)
```

Note: We should be suspicious if no mitochondrial genes show up in the dataset. Another way to check: search for a known mitochondrial gene, e.g. ATP8 (also called MT-ATP8). In this dataset, the name doesn't contain "MT". However, the correct feature (ENSEMBL ID ENSG00000228253) is present in our annotation.

**Annotation problems in general are very common** and should be always considered carefully.

Most modern annotations, e.g. ones used by Cell Ranger, will have mitochondrial genes names that start with MT. Here we're using `org.Hs.eg.db`, which also doesn't support chromosomes (so we don't know where the genes are located).

```{r}
columns(org.Hs.eg.db)
```

Let's try a different database.

```{r}
ensdb_genes <- genes(EnsDb.Hsapiens.v86)
MT_names <- ensdb_genes[seqnames(ensdb_genes) == "MT"]$gene_id
is_mito <- rownames(umi) %in% MT_names
table(is_mito)
```

Important QC metrics: **per-cell, per-gene**. Popular per-cell metrics: total number of counts (UMIs), total number of detected genes, total number of mitochondrial counts, percent of mitochondrial counts.

```{r}
umi_cell <- perCellQCMetrics(umi,subsets=list(Mito=is_mito))
umi_feature <- perFeatureQCMetrics(umi)
head(umi_cell)
```

```{r}
head(umi_feature)
```

Add calculated metrics above to per-cell and per-gene metadata.

```{r}
umi <- addPerCellQC(umi, subsets=list(Mito=is_mito))
umi <- addPerFeatureQC(umi)
```

Set thresholds for what we consider "high enough quality" genes and cells for downstream analysis. This is done manually; we should take into account the distribution of the dataset.

```{r}
hist(
    umi$total,
    breaks = 100
)
abline(v = 25000, col = "red")
```

```{r}
hist(
  umi_cell$detected,
  breaks = 100
)
abline(v = 7000, col = "red")
```

A common strategy is to filter out points that are above 3 median absolute deviations (MAD) in *any* of the QC metrics. **Hallmarks of low-quality cells: low *number* of detected genes, high *percentage* of MT genes.**

```{r}
qc.lib2 <- isOutlier(umi_cell$sum, log=TRUE, type="lower")
attr(qc.lib2, "thresholds")
```

```{r}
qc.nexprs2 <- isOutlier(umi_cell$detected, log=TRUE, type="lower")
attr(qc.nexprs2, "thresholds")
```

```{r}
qc.spike2 <- isOutlier(umi_cell$altexps_ERCC_percent, type="higher")
attr(qc.spike2, "thresholds")
```

```{r}
qc.mito2 <- isOutlier(umi_cell$subsets_Mito_percent, type="higher")
attr(qc.mito2, "thresholds")
```

```{r}
discard2 <- qc.lib2 | qc.nexprs2 | qc.spike2 | qc.mito2
DataFrame(LibSize=sum(qc.lib2), NExprs=sum(qc.nexprs2), SpikeProp=sum(qc.spike2), MitoProp=sum(qc.mito2), Total=sum(discard2))
```

All the actions performed above could be done in one scater command, `quickPerCellQC`:

```{r}
reasons <- quickPerCellQC(umi_cell, sub.fields=c("subsets_Mito_percent", "altexps_ERCC_percent"))
colSums(as.matrix(reasons))
```

Adding a metadata column to keep info on whether a cell is discarded or not:

```{r}
umi$discard <- reasons$discard
```

Never underestimate the usefulness of plotting for QC! Cells with low UMI counts and high % of mitochondrial content are dead or dying.

```{r}
plotColData(umi, x="sum", y="subsets_Mito_percent", colour_by="discard")
```

```{r}
plotColData(umi, x="sum", y="detected", colour_by="discard")
```

```{r}
plotColData(umi, x="altexps_ERCC_percent", y="subsets_Mito_percent",colour_by="discard")
```

Split by batch to see if there are any batch effects (particularly important for RNAseq). (Reminder: the data are from sequencing of iPSCs from 3 individuals.)

```{r}
suppressPackageStartupMessages(library(scales))
plotColData(umi, x="sum", y="detected", colour_by="discard", other_fields = "individual") + 
  facet_wrap(~individual) + scale_x_continuous(labels = unit_format(unit = "k", scale = 1e-3))
```

```{r}
plotColData(umi, x="sum", y="detected", colour_by="discard", other_fields = "replicate") + 
  facet_wrap(~replicate)  + scale_x_continuous(labels = unit_format(unit = "k", scale = 1e-3))
```

Now we start having fun! Let's look at highly expressed genes.

```{r}
plotHighestExprs(umi, exprs_values = "counts", 
                 feature_names_to_plot = "SYMBOL", colour_cells_by="detected")
```

Keep genes that are detected (expression value \> 1) by at least 2 cells.

```{r}
keep_feature <- nexprs(umi,byrow = TRUE,detection_limit = 1) >= 2
rowData(umi)$discard <- ! keep_feature
table(rowData(umi)$discard)
```

Make a new assay: `logcounts_raw` using log2-transformed counts with pseudocount 1, because we don't want to break the machine by making it calculate log2(0).

```{r}
assay(umi, "logcounts_raw") <- log2(counts(umi) + 1)
```

Save it to the `tung` folder for further analysis.

```{r}
saveRDS(umi, file = "course/data/tung/umi.rds")
```

### Data Visualization and Dimensionality Reduction

Create another `SingleCellExperiment` where poorly expressed genes and low quality cells are removed.

```{r}
umi.qc <- umi[! rowData(umi)$discard,! colData(umi)$discard]
```

If we run PCA without transformation or normalization, we mostly see the effects of sequencing depth, i.e., cells with lots of expression and especially highly expressed genes dominate the principal components.

```{r}
umi <- runPCA(umi, exprs_values = "counts")
dim(reducedDim(umi, "PCA"))
```

```{r}
plotPCA(umi, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

Re-run PCA with log-transformed counts.

```{r}
umi <- runPCA(umi, exprs_values = "logcounts_raw")
dim(reducedDim(umi, "PCA"))
```

```{r}
plotPCA(umi, colour_by = "batch", size_by = "detected", shape_by = "individual")
```
Note: **Log-transformation is NOT ENOUGH to account for different technical factors between the cells, e.g., sequencing depth. Use `logcounts` instead of `logcounts_raw`, because `logcounts` are log-transformed AND normalized by library size (e.g., counts per million).** Keep in mind we haven't filtered for cells that may be doublets either. We're using logcounts_raw for demo only. 

Now we'll do PCA on the umi.qc dataframe:

```{r}
umi.qc <- runPCA(umi.qc, exprs_values = "logcounts_raw")
dim(reducedDim(umi.qc, "PCA"))
```

```{r}
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

Only the top 500 *most variable genes* are used to calculate the PCA by the `scater` package. We can adjust this.

PCA when we use all detected genes:

```{r}
umi.qc <- runPCA(umi.qc, exprs_values = "logcounts_raw", ntop=nrow(umi.qc))
dim(reducedDim(umi.qc, "PCA"))
```

```{r}
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

PCA when we use only the top 50 genes
```{r}
umi.qc <- runPCA(umi.qc, exprs_values = "logcounts_raw", ntop=50)
dim(reducedDim(umi.qc, "PCA"))
```

```{r}
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

**t-distributed Stochastic Neighbor Embedding (tSNE)** combines dimensionality reduction with random walks on nearest-neighbor network. Because this is a *stochastic algorithm*, running this method multiple times on the same dataset will give us different plots. For reproducibility, set the seed for the random-number generator. 

[More on using tSNE plots.](https://distill.pub/2016/misread-tsne/)
[The art of using t-SNE for single-cell transcriptomics](https://www.nature.com/articles/s41467-019-13056-x)

What it looks like without QC:

```{r}
set.seed(123456)
umi <- runTSNE(umi, exprs_values = "logcounts_raw", perplexity = 130)
plotTSNE(umi, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

What it looks like post-QC:

```{r}
set.seed(123456)
umi.qc <- runTSNE(umi.qc, exprs_values = "logcounts_raw", perplexity = 130)
plotTSNE(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

Based on these plots, I am inclined to think there are significant differences between individuals in terms of transcriptomic profiles of their iPSCs. But it's hard to interpret these without any external references/ground truths (for example, stem cell markers like OCT4, NANOG, SOX2).

t-SNE plot when perplexity is changed to 10.

```{r}
set.seed(123456)
umi.qc <- runTSNE(umi.qc, exprs_values = "logcounts_raw", perplexity = 10)
plotTSNE(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

t-SNE plot when perplexity is changed to 200.

```{r}
set.seed(123456)
umi.qc <- runTSNE(umi.qc, exprs_values = "logcounts_raw", perplexity = 200)
plotTSNE(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

### Confounding Factors

Challenge: it's hard to have a *true* technical replicate to distinguish *biological* and *technical* variability.

```{r}
umi.qc <- runPCA(umi.qc, exprs_values = "logcounts_raw")
dim(reducedDim(umi.qc, "PCA"))
```

```{r}
plotPCA(umi.qc, colour_by = "batch", size_by = "sum", shape_by = "individual")
```

With `scater`, we can test whether experimental and QC variables correlate with any of the principal components.

```{r}
logcounts(umi.qc) <- assay(umi.qc, "logcounts_raw")
getExplanatoryPCs(umi.qc, variables = "sum")
```

```{r}
plotExplanatoryPCs(umi.qc, variables = "sum")
```

```{r}
logcounts(umi.qc) <- NULL
```

PC1 is almost completely (86%) explained by the total UMI counts (sequencing depth).

Explanatory variables

```{r}
plotExplanatoryVariables(umi.qc,exprs_values = "logcounts_raw", 
                         variables = c("detected", "sum", "batch", "individual", "altexps_ERCC_percent", "subsets_Mito_percent"))
```

This confirms/reemphasized the impact of the number of genes detected and number of counts, hence these can be conditioned out in a normalization step, or making sure they're accounted for in downstream statistical models. The plot above also shows that the batch explains more than the individual. (Remember the earlier note about how it's hard to get *true* technical replicates.)

More confounders to consider: are cells dying? are cells replicating? what sequencing protocols were used, and what was the coverage like? was there a bias of more A/T nucleotides being used vs. G/C in the transcripts? 

A method for identifying and subtracting cell cycle or apoptosis: [scLVM](https://github.com/PMBio/scLVM)

### Normalization

`scater` has `normaliseExprs()` that takes the expression matrix and a design matrix. 

Size-factor normalization: correct for library size, remove effects of some confounders and explanatory variables.

Some, not all, quantification methods incorporate library size when determining gene expression, e.g., `Cufflinks`, `RSEM`. For these, we don't need to add a step to normalize for libary size.

Normalization factor: estimate of the library size relative to the other cells. Some that work for bulk RNA-seq can also work for scRNA-seq, such as upper quartile (UQ), CPM, RPKM, FPKM, TPM.

More ways to normalize: relative log expression (RLE), trimmed mean of M-values (TMM) (assign one cell as a reference, assumption: most genes are not differentially expressed), `scran` (CPM specialized for single cell by pooling cells to calculate a normalization factor), downsampling (usually need to downsample datasets multiple rounds). 

```{r}
suppressPackageStartupMessages(library(scRNA.seq.funcs))
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(scran))

set.seed(1234567)
umi <- readRDS("course/data/tung/umi.rds")
umi.qc <- umi[! rowData(umi)$discard, ! colData(umi)$discard]
```

PCA on logcounts_raw

```{r}
umi.qc <- runPCA(umi.qc, exprs_values = "logcounts_raw")
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

PCA on CPM-normalized data

```{r}
logcounts(umi.qc) <- log2(calculateCPM(umi.qc) + 1)
umi.qc <- runPCA(umi.qc)
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

RLE plots can be useful to assess whether normalization was successful.

```{r}
plotRLE(umi.qc, exprs_values = "logcounts_raw", colour_by = "batch") + ggtitle("RLE plot for logcounts_raw")
```
```{r}
plotRLE(umi.qc, exprs_values = "logcounts", colour_by = "batch") + ggtitle("RLE plot for log2(CPM) counts")
```

PCA on `scran`-normalized data.

`scran` uses clustering to make normalization, also referred to as "normalization by deconvolution".

```{r}
qclust <- quickCluster(umi.qc, min.size = 30)
table(qclust)
```

Compute size factors using clusters.

```{r}
umi.qc <- computeSumFactors(umi.qc, clusters = qclust)
```

```{r}
umi.qc <- logNormCounts(umi.qc)
```

```{r}
umi.qc <- runPCA(umi.qc)
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

```{r}
plotRLE(umi.qc, exprs_values = "logcounts", colour_by = "batch")
```

check the size factors computed by `scran`

```{r}
summary(sizeFactors(umi.qc))
```

PCA with downsampled data 

```{r}
logcounts(umi.qc) <- log2(Down_Sample_Matrix(counts(umi.qc)) + 1)
umi.qc <- runPCA(umi.qc)
plotPCA(umi.qc, colour_by = "batch", size_by = "detected", shape_by = "individual")
```

```{r}
plotRLE(umi.qc, exprs_values = "logcounts", colour_by = "batch")
```

### Dealing with confounders 

**BATCH CORRECTION.**

Two scenarios for scRNA-seq: 
(1) expect cell composition to be the same, methods developed for bulk RNA-seq perform well. often true for biological replicates in the same experiment.
(2) overlap between datasets is partial, e.g., if dataset has healthy and disease tissue, which can differ in cell type composition. for this, mutual nearest neighbor (MNN)-based methods tend to perform better.

```{r}
# suppressPackageStartupMessages(library(scRNA.seq.funcs))
# suppressPackageStartupMessages(library(scater))
# suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(sva))
suppressPackageStartupMessages(library(batchelor))
suppressPackageStartupMessages(library(kBET))

set.seed(1234567)
```

Load and normalize the Tung dataset.

```{r}
umi <- readRDS("course/data/tung/umi.rds")
umi.qc <- umi[! rowData(umi)$discard, ! colData(umi)$discard]
qclust <- quickCluster(umi.qc, min.size = 30)
umi.qc <- computeSumFactors(umi.qc, clusters = qclust)
```

```{r}
umi.qc <- logNormCounts(umi.qc)
```

Using `ComBat` to eliminate batch effects.

```{r}
assay(umi.qc, "combat") <- ComBat(logcounts(umi.qc), batch = umi.qc$replicate)
```
ComBat correction accounting for total features as a co-variate. (from `sva`)

```{r}
assay(umi.qc, "combat_tf") <- ComBat(logcounts(umi.qc), batch = umi.qc$detected)
```
MNN (`batchelor`)

```{r}
mnn_out <- fastMNN(umi.qc, batch = umi.qc$replicate)
assay(umi.qc, "mnn") <- assay(mnn_out, "reconstructed")
```

**Evaluate and compare batch-removal approaches**

Which method is most effective? Tricky because of what is a technical confounder vs. what is interesting biological variability. At the end of the day, it depends on the biological question we're asking. 

Evaluate effectiveness of normalization methods by inspecting the PCA plot.

```{r}
for(n in assayNames(umi.qc)) {
  tmp <- runPCA(umi.qc, exprs_values = n, ncomponents = 20)
  
  print(
    plotPCA(
      tmp,
      colour_by = "batch",
      size_by = "detected",
      shape_by = "individual"
    ) +
    ggtitle(n)
  )
}
```

Can also examine effectiveness with RLE.

```{r}
res <- list()
for(n in assayNames(umi.qc)) {
  res[[n]] <- suppressWarnings(calc_cell_RLE(assay(umi.qc, n)))
}
par(mar=c(6, 4, 1, 1))
boxplot(res, las=2)
```

Examine effectiveness by considering intermingling of points from different batches in local subsamples of the data. If there are no batch-effects: proportion of cells from each batch in any local region should be equal to the global proportion of cells in each batch. 

`kBET` can only be applied if a perfectly balanced design has been used. 

```{r}
compare_kBET_results <- function(sce){
  sce <- umi.qc
  indiv <- unique(as.character(sce$individual))
  norms <- assayNames(sce) # get all the normalizations
  results <- list()
  for (i in indiv){
    for (j in norms){
      tmp <- kBET(
        df = t(assay(sce[, sce$individual == i], j)),
        batch = sce$batch[sce$individual == i],
        heuristic = TRUE,
        verbose = FALSE,
        addTest = FALSE,
        plot = FALSE
      )
      results[[i]][[j]] <- tmp$summary$kBET.observed[1]
    }
  }
  return(do.call(rbind.data.frame, results))
}

eff_debatching <- compare_kBET_results(umi.qc)
eff_debatching
```

Visualize the output of kBET calculation.

```{r}
suppressPackageStartupMessages(library("reshape2"))
suppressPackageStartupMessages(library("RColorBrewer"))

dod <- melt(as.matrix(eff_debatching), value.name = "kBET")
colnames(dod)[1:2] <- c("Normalization", "Individual")

colorset <- c("gray", brewer.pal(n = 9, "RdYlBu"))

ggplot(dod, aes(Normalization, Individual, fill=kBET)) +
  geom_tile() +
  scale_fill_gradient2(
    na.value = "gray",
    low = colorset[2],
    mid = colorset[6],
    high = colorset[10],
    midpoint = 0.5, limit = c(0,1)
  ) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  theme(
    axis.text.x = element_text(
      angle = 45,
      vjust = 1,
      size = 12,
      hjust = 1
    )
  ) +
  ggtitle("Effect of batch regression methods per individual")
```

### Up next: Biological Analysis